internal/backend/local/test.go

```diff
@@ -628,7 +610,46 @@ func (runner *TestFileRunner) run(run *moduletest.Run, file *moduletest.File, st
 	// our prior run outputs so future run blocks can access it.
 	runner.EvalContext.SetOutput(run, outputVals)
 
-	return updated, true
+	// Only update the most recent run and state if the state was
+	// actually updated by this change. We want to use the run that
+	// most recently updated the tracked state as the cleanup
+	// configuration.
+	runner.EvalContext.SetFileState(key, &graph.TestFileState{
+		Run:   run,
+		State: updated,
+	})
+}
+
+// checkForMissingExpectedFailures checks for missing expected failures in the diagnostics.
+// It updates the run status based on the presence of errors or missing expected failures.
+func (runner *TestFileRunner) checkForMissingExpectedFailures(run *moduletest.Run, diags tfdiags.Diagnostics) (failOrErr bool) {
+	// If there are errors in the diagnostics, update the run status to error.
+	if diags.HasErrors() {
+		run.Status = moduletest.Error
+	}
+
+	// Retrieve diagnostics that are either unrelated to expected failures or report missing expected failures.
+	unexpectedDiags := run.ValidateExpectedFailures(diags)
+	var nonFailureDiags tfdiags.Diagnostics
+	for _, diag := range unexpectedDiags {
+		switch {
+		// If any diagnostic indicates a missing expected failure, update the run status to fail.
+		case diag.Description().Summary == moduletest.MissingFailureSummary:
```

Comment:
I think the implementation here is fine, but we should use the `Extra` functionality of the diagnostic package instead of matching against the summary text. This just means we're being much more precise.

You could make an interface in the `run` package:

```

type DiagnosticExtraFromMissingExpectedFailure interface {
  DiagnosticFromMissingExpectedFailure() bool
}

func DiagnosticFromMissingExpectedFailure(diag tfdiags.Diagnostic) bool {
	maybe := tfdiags.ExtraInfo[DiagnosticExtraFromMissingExpectedFailure](diag)
	if maybe == nil {
		return false
	}
	return maybe.DiagnosticFromMissingExpectedFailure()
}

```

Then set the extra of the diagnostic to be an implementation of the interface. Then here you can call `moduletest.DiagnosticFromMissingExpectedFailure(diag)`.

Reply:
Addressed in [d0108ed](https://github.com/hashicorp/terraform/pull/36447/commits/d0108eda1d7de1c6f8b6f22a92733f94755e1013)